walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,53 +15,122 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.LinkedList;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Stack;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     /**
-     * importantRanges - store time ranges that are important for at least one attendee in the request
-     * busyIntervals - eventually store merged important time ranges
-     * freeIntervals - used to collect intevals that are between busyIntervals   
+     * importantRanges - store time ranges that are important for at least one mandatory attendee in the request (doesn't apply to opitonal attendees)
+     * optionalImportantRanges - store time ranges that are important for mandatory as well as optional attendees 
+     * busyIntervals - eventualy store merged important time ranges
+     * optionalBusyRanges - same as above but for mandatory and optional attendees
+     * freeIntervals - used to collect intevals that are between busyIntervals
+     * freeRangesWithOptionals - same as above but for mandatory and optional attendees   
      */
     ArrayList<TimeRange> importantRanges = new ArrayList<TimeRange>();
-    Stack<TimeRange> busyIntervals = new Stack<TimeRange>();
-    ArrayList<TimeRange> freeIntervals = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> optionalImportantRanges = new ArrayList<TimeRange>();
+    Stack<TimeRange> busyRanges = new Stack<TimeRange>();
+    Stack<TimeRange> optionalBusyRanges = new Stack<TimeRange>();
+    ArrayList<TimeRange> freeRanges = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> freeRangesWithOptionals = new ArrayList<TimeRange>();
```

Comment:
These two variables are not used until the very end. Don't prematurely define variables. Only define variables before you need them. It reduces mental burden & complexity when people read your code.

Reply:
If you're interested, have a quick read at page 246 (http://aroma.vn/web/wp-content/uploads/2016/11/code-complete-2nd-edition-v413hav.pdf)

Reply:
Seems like great resource, thanks.